enterprise/server/byte_stream_server_proxy/byte_stream_server_proxy.go

```diff
@@ -145,221 +256,180 @@ func recordReadMetrics(cacheStatus string, proxyRequestType string, err error, b
-		}
+	stream = nil // use meteredStream
+	var err error
+	if authutil.EncryptionEnabled(ctx, s.authenticator) {
+		err = s.writeRemoteOnly(ctx, meteredStream)
+	} else if proxy_util.SkipRemote(ctx) {
+		err = s.writeLocalOnly(meteredStream)
+	} else {
+		err = s.dualWrite(ctx, meteredStream)
 	}
+	recordWriteMetrics(meteredStream.GetByteCount(), err, requestTypeLabel)
+	return err
+}
 
-	bytesRead := 0
+func (s *ByteStreamServerProxy) writeRemoteOnly(ctx context.Context, stream bspb.ByteStream_WriteServer) error {
+	remoteStream, err := s.remote.Write(ctx)
+	if err != nil {
+		return err
+	}
 	for {
-		rsp, err := remoteReadStream.Recv()
-		if rsp != nil {
-			bytesRead += len(rsp.Data)
+		req, err := stream.Recv()
+		if err == io.EOF {
+			break
 		}
 		if err != nil {
-			if err == io.EOF {
-				if err := localWriteStream.commit(); err != nil {
-					log.CtxInfof(ctx, "error committing local write: %s", err)
-				}
-				break
-			}
-			log.CtxInfof(ctx, "error streaming from remote for read through: %s", err)
-			return bytesRead, err
+			return err
 		}
-
-		if err := localWriteStream.send(rsp.Data); err != nil {
-			log.CtxInfof(ctx, "Error writing locally for read through: %s", err)
-			localWriteStream = &discardingLocalWriter{}
+		if err = remoteStream.Send(req); err != nil {
+			return err
 		}
-		if err = stream.Send(rsp); err != nil {
-			return bytesRead, err
+		if req.GetFinishWrite() {
+			break
 		}
 	}
-	return bytesRead, nil
+	resp, err := remoteStream.CloseAndRecv()
+	if err != nil {
+		return err
+	}
+	return stream.SendAndClose(resp)
 }
 
-func (s *ByteStreamServerProxy) Write(stream bspb.ByteStream_WriteServer) error {
-	resp, err := s.write(stream)
-	requestTypeLabel := proxy_util.RequestTypeLabelFromContext(stream.Context())
-	recordWriteMetrics(resp, err, requestTypeLabel)
-	return err
+func (s *ByteStreamServerProxy) writeLocalOnly(stream bspb.ByteStream_WriteServer) error {
+	return s.local.Write(stream)
 }
 
-func (s *ByteStreamServerProxy) write(stream bspb.ByteStream_WriteServer) (*bspb.WriteResponse, error) {
-	ctx, spn := tracing.StartSpan(stream.Context())
-	defer spn.End()
+// TODO(iain): investigate performance of making the local write async
+func (s *ByteStreamServerProxy) dualWrite(ctx context.Context, stream bspb.ByteStream_WriteServer) error {
+	// Grab the first frame from the client so the local writer can be created
+	req, err := stream.Recv()
+	if err != nil {
+		return err
+	}
 
-	var local bspb.ByteStream_WriteClient
-	if !authutil.EncryptionEnabled(ctx, s.authenticator) {
-		localWriter, err := s.local.Write(ctx)
-		if err != nil {
-			log.CtxInfof(ctx, "error opening local bytestream write stream for write: %s", err)
-			localWriter = nil
-		}
-		local = localWriter
+	localWriteStream, err := s.local.BeginWrite(ctx, req)
+	if err == nil {
+		defer localWriteStream.Close()
+	} else {
+		log.CtxDebugf(ctx, "Error opening local write stream: %v", err)
```

Comment:
Should this be a higher log level? Seems like this shouldn't happen

Reply:
Yeah, good call, done